### PR TITLE
merge develop into main

### PR DESCRIPTION
**Pull request by TomasDmArg**
## Changelog
- Externalized ABIs
- Externalized environment variables
- Improved NFTs
- Transfer refactor
- Deposits listening using subgraph (pending to be called through GCP Cloud Scheduler)
- Improved responses / error handling
- Fixed token authentication with secret manager
- Transfers with User Operations working

### Closes the following issues:

- #22 
- #63
- #19
- #72
- #26
- #21
- #32

### CODE REVIEW

**Comment by dappsar** [comment]
@TomasDmArg   completar descripción con litstado de IDs de issues que se están desplegando.

**Comment by TomasDmArg** [comment]
> @TomasDmArg completar descripción con litstado de IDs de issues que se están desplegando.

Agregadas

**Comment by dappsar** [comment]
deprecated

**Review by dappsar — APPROVED** [review]
review items fixed.

File: scripts/docker-build.sh [hunk 1]
```diff
@@ -24,7 +24,11 @@ docker build \
     --build-arg ICP_MNEMONIC="$ICP_MNEMONIC" \
     --build-arg BOT_API_URL="$BOT_API_URL" \
     --build-arg BOT_DATA_TOKEN="$BOT_DATA_TOKEN" \
+    --build-arg ARBITRUM_SEPOLIA_RPC_URL="$ARBITRUM_SEPOLIA_RPC_URL" \
     --build-arg NFT_UPLOAD_IMAGE_ICP="$NFT_UPLOAD_IMAGE_ICP" \
     --build-arg NFT_UPLOAD_IMAGE_IPFS="$NFT_UPLOAD_IMAGE_IPFS" \
+    --build-arg GCP_BUCKET_BASE_URL="$GCP_BUCKET_BASE_URL" \
+    --build-arg CHATIZALO_TOKEN="$CHATIZALO_TOKEN" \
+    --build-arg FRONTEND_TOKEN="$FRONTEND_TOKEN" \
```
[the remark this answers is not in the document]
Comment (dappsar) [reply scripts/docker-build.sh:32]: - Fixed: Added token header in frontend code and env variables.

File: cloudbuild.yaml [hunk 2]
```diff
@@ -25,8 +19,9 @@ steps:
           - BOT_DATA_TOKEN=${_BOT_DATA_TOKEN}
           - NFT_UPLOAD_IMAGE_ICP=${_NFT_UPLOAD_IMAGE_ICP}
           - NFT_UPLOAD_IMAGE_IPFS=${_NFT_UPLOAD_IMAGE_IPFS}
+          - GCP_BUCKET_BASE_URL=${_GCP_BUCKET_BASE_URL}
+      secretEnv: ['CHATIZALO_TOKEN', 'FRONTEND_TOKEN']
```
[the remark this answers is not in the document]
Comment (dappsar) [reply cloudbuild.yaml]: Fixed in commit #https://github.com/P4-Games/ChatterPay-Backend/commit/7bbb837f9e36c297e0ce0f568973332d36af2901

File: src/middleware/authMiddleware.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * 2. Checks if the header exists and starts with 'Bearer '.
+ * 3. Verifies the token using the `verifyToken` function.
+ * 4. If the token is valid, adds the token type to the request headers.
+ * 5. If any step fails, it sends a 401 response with an error message.
+ */
+export async function authMiddleware(request: FastifyRequest, reply: FastifyReply): Promise<void> {
+  const authHeader: string | undefined = request.headers.authorization;
+
+  if (!authHeader || !authHeader.startsWith('Bearer ')) {
+    reply.code(401).send({ error: 'No se proporcionó token de autenticación' });
```
[the remark this answers is not in the document]
Comment (dappsar) [reply src/middleware/authMiddleware.ts:31]: fixed in commit #https://github.com/P4-Games/ChatterPay-Backend/commit/7bbb837f9e36c297e0ce0f568973332d36af2901

File: cloudbuild.yaml [hunk 4]
```diff
@@ -69,10 +64,18 @@ steps:
               'NFT_UPLOAD_IMAGE_ICP=${_NFT_UPLOAD_IMAGE_ICP}',
               '--build-arg',
               'NFT_UPLOAD_IMAGE_IPFS=${_NFT_UPLOAD_IMAGE_IPFS}',
+              '--build-arg',
+              'GCP_BUCKET_BASE_URL=${_GCP_BUCKET_BASE_URL}',
+              '--build-arg',
+              'ARBITRUM_SEPOLIA_RPC_URL=${_ARBITRUM_SEPOLIA_RPC_URL}',
+              '--build-arg',
+              'CHATIZALO_TOKEN=$$CHATIZALO_TOKEN',
```
[the remark this answers is not in the document]
Comment (dappsar) [reply cloudbuild.yaml]: Fixed in commit #https://github.com/P4-Games/ChatterPay-Backend/commit/7bbb837f9e36c297e0ce0f568973332d36af2901

File: Dockerfile [hunk 5]
```diff
@@ -37,6 +41,10 @@ ENV BOT_API_URL $BOT_API_URL
 ENV BOT_DATA_TOKEN $BOT_DATA_TOKEN
 ENV NFT_UPLOAD_IMAGE_ICP $NFT_UPLOAD_IMAGE_ICP
 ENV NFT_UPLOAD_IMAGE_IPFS $NFT_UPLOAD_IMAGE_IPFS
+ENV GCP_BUCKET_BASE_URL $GCP_BUCKET_BASE_URL
+ENV ARBITRUM_SEPOLIA_RPC_URL $ARBITRUM_SEPOLIA_RPC_URL
+ENV CHATIZALO_TOKEN $CHATIZALO_TOKEN
```
[the remark this answers is not in the document]
Comment (dappsar) [reply Dockerfile:46]: Fixed in commit #https://github.com/P4-Games/ChatterPay-Backend/pull/140/commits/89f85cc83efc59c95e8925cc15333d3934b941f6

**Pull request merged**